### PR TITLE
Added actor critic example

### PR DESCRIPTION
**Pull request by apoorvnandan**
A simple implementation of actor-critic method in CartPole environment.

- Solves the task in ~350 episodes. (3-4 minutes on CPU)
- Policy network implemented using Functional API. Initial layer shared between actor and critic.
- Network weights updated using `tf.GradientTape`
- Prints running reward every 10 episodes.

### CODE REVIEW

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+"""
+
+import gym
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow import math
+from tensorflow.keras import losses
+from tensorflow.keras import layers
+from tensorflow.keras import optimizers
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:19]: You don't need these imports, just use `tf.` and `keras.` for thins like `tf.math`, `keras.losses`, etc.

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+seed = 543
+gamma = 0.99
+env = gym.make("CartPole-v0")
+env.seed(seed)
+eps = np.finfo(np.float32).eps.item()
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:25]: This deserves a comment

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+critic_value_history = []
+rewards_history = []
+running_reward = 0
+episode_count = 0
+
+while True:  # run till solved
+    state = env.reset()
+    episode_reward = 0
+    with tf.GradientTape() as tape:  # keep track of ops for backprop
+        for timestep in range(1, 10000):
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:57]: This should be configurable

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+action_probs_history = []
+critic_value_history = []
+rewards_history = []
+running_reward = 0
+episode_count = 0
+
+while True:  # run till solved
+    state = env.reset()
+    episode_reward = 0
+    with tf.GradientTape() as tape:  # keep track of ops for backprop
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:56]: This comment is not needed

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        for timestep in range(1, 10000):
+            state = tf.convert_to_tensor(state)
+            state = tf.expand_dims(state, 0)
+
+            action_probs, critic_value = model(state)
+            critic_value_history.append(critic_value[0, 0])
+
+            # Sample action from action prob distribution
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:64]: Avoid abbreviations, spell out word in full

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        running_reward = 0.05 * episode_reward + (1 - 0.05) * running_reward
+
+        # Calculate expected value from rewards
+        returns = []
+        discounted_sum = 0
+        for r in rewards_history[::-1]:
+            discounted_sum = r + gamma * discounted_sum
+            returns.insert(0, discounted_sum)
+
+        # Normalise
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:85]: Normalize

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        for log_prob, value, ret in history:
+            diff = ret - value
+            actor_losses.append(-log_prob * diff)  # actor loss
+            critic_losses.append(
+                losses.Huber()(tf.expand_dims(value, 0), tf.expand_dims(diff, 0))
+            )  # critic loss
+
+        # Backprop
+        loss_value = sum(actor_losses) + sum(critic_losses)
+
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:102]: Remove blank line (otherwise the comment seems attached to the loss sum above)

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        returns = returns.tolist()
+
+        history = zip(action_probs_history, critic_value_history, returns)
+        actor_losses = []
+        critic_losses = []
+        for log_prob, value, ret in history:
+            diff = ret - value
+            actor_losses.append(-log_prob * diff)  # actor loss
+            critic_losses.append(
+                losses.Huber()(tf.expand_dims(value, 0), tf.expand_dims(diff, 0))
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:97]: You don't need to reinstantiate `Huber` at every step, either instantiate it once at the start or use `keras.losses.huber` (which returns one entry per sample)

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+from tensorflow.keras import optimizers
+
+seed = 543
+gamma = 0.99
+env = gym.make("CartPole-v0")
+env.seed(seed)
+eps = np.finfo(np.float32).eps.item()
+
+"""
+## Implement Actor Critic network
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:28]: Please add an intro section explaining what this script does, why it matters, and provide a link to the reference paper(s).

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+"""
+
+optimizer = optimizers.Adam(learning_rate=0.01)
+action_probs_history = []
+critic_value_history = []
+rewards_history = []
+running_reward = 0
+episode_count = 0
+
+while True:  # run till solved
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:53]: The various steps in this loop should be more abundantly commented; explain your full though process

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+    # Log details
+    episode_count += 1
+    if episode_count % 10 == 0:
+        template = "running reward: {:.2f} at episode {}"
+        print(template.format(running_reward, episode_count))
+
+    if running_reward > 195:  # Condition to consider the task solved
+        print("Solved at episode {}!".format(episode_count))
+        break
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:119]: Consider adding some kind of visualization at the end, if possible

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
+"""
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:1]: Also, we should name the folder `rl` (like `nlp`).

File: examples/reinforcement_learning/actor_critic_cartpole.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+while True:  # run till solved
+    state = env.reset()
+    episode_reward = 0
+    with tf.GradientTape() as tape:  # keep track of ops for backprop
+        for timestep in range(1, 10000):
+            state = tf.convert_to_tensor(state)
+            state = tf.expand_dims(state, 0)
+
+            action_probs, critic_value = model(state)
```
Comment (fchollet) [inline examples/reinforcement_learning/actor_critic_cartpole.py:61]: Also it would be more performant to put the 3 lines above in a `tf.function`

**Comment by apoorvnandan** [comment]
Thank you for the comments. I'll update the code.

File: examples/reinforcement_learning/actor_critic_cartpole.py (hunk 1 above)
Comment (apoorvnandan) [reply examples/reinforcement_learning/actor_critic_cartpole.py:19]: Got it.

I was actually wondering about the exact best practice in such cases. 
If only a single optimizer and loss function are being instantiated, we use `keras.losses.Huber` or `keras.optimizers.Adam` and if multiple things from a submodule are being used (like multiple layers), we import `from tf.keras import layers` and use `layers.Dense` etc. Is this understanding correct?

File: examples/reinforcement_learning/actor_critic_cartpole.py (hunk 11 above)
Comment (apoorvnandan) [reply examples/reinforcement_learning/actor_critic_cartpole.py:119]: I can plot the rewards for starters. Can videos be embedded in tutobooks?

**Comment by apoorvnandan** [comment]
Added a bunch of annotations throughout the code. 
Added an introduction section as well. 
Please take a look.

Regarding the visualization, calling `env.render()` inside the time loop pops up a little window which shows how the agent attempts to balance the pole on the cart. I'm not sure if this can be inserted inside the tutobook? Or we can add the line and write a comment that a window will pop up with the live show?

File: examples/reinforcement_learning/actor_critic_cartpole.py (hunk 13 above)
Comment (apoorvnandan) [reply examples/reinforcement_learning/actor_critic_cartpole.py]: Tested this. 

@tf.function takes 0.0005 seconds on an average.
The current code takes 0.0009 seconds on an average.

Both these numbers are very low compared to the overall time taken per episode, which varies from 0.01 to 0.06 seconds on an average.

File: examples/reinforcement_learning/actor_critic_cartpole.py (hunk 11 above)
Comment (fchollet) [reply examples/reinforcement_learning/actor_critic_cartpole.py]: Ultimately we need to be able to display the examples as a webpage. How difficult would it be to create a gif?

If easy, do it in the notebook and display the gif in-line. If difficult, then:
- Create it offline
- Include it in your PR by putting it in `site/img/examples_extra/`
- In the tutobook text, including the gif in the text via markdown tags
- Also in the tutobook text, show the code the user would normally use to display the animation

File: examples/reinforcement_learning/actor_critic_cartpole.py (hunk 13 above)
Comment (fchollet) [reply examples/reinforcement_learning/actor_critic_cartpole.py]: Ok, we can leave it as is then.

File: examples/reinforcement_learning/actor_critic_cartpole.py (hunk 11 above)
Comment (apoorvnandan) [reply examples/reinforcement_learning/actor_critic_cartpole.py]: Let me check about the gifs.

Comment (apoorvnandan) [reply examples/reinforcement_learning/actor_critic_cartpole.py]: Created the gifs. By `site/img/examples_extra/` do you mean I should create a directory called `site` in the root directory of this repo?

Comment (fchollet) [reply examples/reinforcement_learning/actor_critic_cartpole.py]: Here's an even simpler alternative: host the image externally, e.g. on imgur, and copy the link. Or else, just put it in `examples/rl/actor_critic_cartpole/img/` and I'll take care of hosting it appropriately.

Comment (apoorvnandan) [reply examples/reinforcement_learning/actor_critic_cartpole.py]: Okay.

**Comment by apoorvnandan** [comment]
Added visualizations. Please check.

**Comment by apoorvnandan** [comment]
Added the generated files.

File: scripts/.ipynb_checkpoints/tutobooks-checkpoint.py [hunk 14]
```diff
@@ -0,0 +1,516 @@
+"""Keras tutobooks implementation.
```
Comment (fchollet) [inline scripts/.ipynb_checkpoints/tutobooks-checkpoint.py:1]: Please remove this file.

Comment (apoorvnandan) [reply scripts/.ipynb_checkpoints/tutobooks-checkpoint.py]: Ah sorry! Got added by mistake.
